compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp

```diff
@@ -4,225 +4,265 @@
+                                 threadBlockShape[kN] / numWarps[kN],
+                                 threadBlockShape[kK] / numWarps[kK]};
+
+  SmallVector<int64_t> instructionShape;
+  getInstructionShape(pipeline, lhsType.cast<ShapedType>().getElementType(),
+                      instructionShape);
+
+  // Number of software pipeline stages/depth.
+  int64_t softwarePipelineDepth = translationInfo.getSoftwarePipelineDepth();
+
+  //
+  // Begin verification for CUDA and Tensor Core pipelines.
+  //
+#if 0  // Will remove once the review is done and CI passes, keeping till then
+       // to make it easier to review CI failures.
+  std::cout << "matmulShape: " << matmulShape[0] << "x" << matmulShape[1] << "x"
+            << matmulShape[2] << std::endl;
+  std::cout << "threadBlockShape: " << threadBlockShape[0] << "x"
+            << threadBlockShape[1] << "x" << threadBlockShape[2] << std::endl;
+  std::cout << "warpShape: " << warpShape[0] << "x" << warpShape[1] << "x"
+            << warpShape[2] << std::endl;
+  std::cout << "instructionShape: " << instructionShape[0] << "x"
+            << instructionShape[1] << "x" << instructionShape[2] << std::endl;
+#endif
+
+  // Verify the total number of threads in a thread block.
+  int totalNumThreads = workgroupSize[0] * workgroupSize[1] * workgroupSize[2];
+
+  if (totalNumThreads > 1024) {
+    return op->emitError("Total number of threads in a thread block ")
+           << totalNumThreads << " exceeds the limit of 1024";
   }
-  // Verify the workgroup.z component should always be 1
-  if (workgroupSize[2] != 1) {
-    return op->emitOpError("expected workgroup z component to be 1 for ")
-           << pipelineName << ", got " << workgroupSize[2];
+
+  // Verify that the number of threads in x-dimension is a multiple of
+  // kWarpSize.
+  if (workgroupSize[kDimX] % kWarpSize != 0) {
+    return op->emitError("Number of threads in x-dimension ")
+           << workgroupSize[kDimX] << " is not a multiple of " << kWarpSize;
   }
 
-  // The second level of tiling = [M / numWarp.y, N / numWarp.x, K].
-  SmallVector<int64_t, 3> secondLevelTileSizes = {
-      firstLevelTileSizes[0] / workgroupSize[1],
-      firstLevelTileSizes[1] / (workgroupSize[0] / kWarpSize),
-      firstLevelTileSizes[2]};
-
-  // Verify the TensorCore size divides the second level tile size
-  SmallVector<int64_t, 3> tensorCoreSize({16, 16, 8});
-  if (secondLevelTileSizes[0] % tensorCoreSize[0] != 0 ||
-      secondLevelTileSizes[1] % tensorCoreSize[1] != 0 ||
-      secondLevelTileSizes[2] % tensorCoreSize[2] != 0) {
-    return op->emitOpError(
-               "tensorcore size doesn't factor into second level tile size "
-               "for ")
-           << pipelineName;
+  // Verify the number of warps in the problem K dimension is 1.
+  if (numWarps[kK] != 1) {
+    return op->emitError("Number of warps in problem K dimension ")
+           << numWarps[kK] << " is not 1";
```

Comment:
this doesn't make sense to me. There is no such thing as numWarps along K.

Reply:
number of warps in the K dim is 1. If we have more than one warp accumulate TileK dim then what we have is Sliced-K kernel equivalent from CUTLASS. 

`Split_K` is dividing the ProblemK across multiple threadblocks, and `Sliced_K` is dividing Tile_K accross multiple warps. We don't have Sliced_K and we don't plan to support it. Also, thing of it as in the picture above green line shows the warp shape, The green lines are only in matrixC and not in matrixA or matrixB.

Reply:
This condition is to check, if the user has given tile config such that it ends up having numWarp > 1, we catch that.